src/shared/components/common/searchable-select.tsx

```diff
@@ -0,0 +1,201 @@
+import classNames from "classnames";
+import {
+  ChangeEvent,
+  Component,
+  createRef,
+  linkEvent,
+  RefObject,
+} from "inferno";
+import { i18n } from "../../i18next";
+import { Choice } from "../../utils";
+import { Icon, Spinner } from "./icon";
+
+interface SearchableSelectProps {
+  id: string;
+  value?: number | string;
+  options: Choice[];
+  onChange?: (option: Choice) => void;
+  onSearch?: (text: string) => void;
+  loading?: boolean;
+}
+
+interface SearchableSelectState {
+  selectedIndex: number;
+  searchText: string;
+  loadingEllipses: string;
+}
+
+function handleSearch(i: SearchableSelect, e: ChangeEvent<HTMLInputElement>) {
+  const { onSearch } = i.props;
+  const searchText = e.target.value;
+
+  if (onSearch) {
+    onSearch(searchText);
+  }
+
+  i.setState({
+    searchText,
+  });
+}
+
+export class SearchableSelect extends Component<
+  SearchableSelectProps,
+  SearchableSelectState
+> {
+  private searchInputRef: RefObject<HTMLInputElement> = createRef();
+  private toggleButtonRef: RefObject<HTMLButtonElement> = createRef();
+  private loadingEllipsesInterval?: NodeJS.Timer = undefined;
+
+  state: SearchableSelectState = {
+    selectedIndex: 0,
+    searchText: "",
+    loadingEllipses: "...",
+  };
+
+  constructor(props: SearchableSelectProps, context: any) {
+    super(props, context);
+
+    this.handleChange = this.handleChange.bind(this);
+    this.focusSearch = this.focusSearch.bind(this);
+
+    if (props.value) {
+      let selectedIndex = props.options.findIndex(
+        ({ value }) => value === props.value
+      );
+
+      if (selectedIndex < 0) {
+        selectedIndex = 0;
+      }
+
+      this.state = {
+        ...this.state,
+        selectedIndex,
+      };
+    }
+  }
+
+  render() {
+    const { id, options, onSearch, loading } = this.props;
+    const { searchText, selectedIndex, loadingEllipses } = this.state;
+
+    return (
+      <div className="dropdown">
+        <button
+          id={id}
+          type="button"
+          className="custom-select text-start"
+          aria-haspopup="listbox"
+          data-bs-toggle="dropdown"
```

Comment:
Not a huge fan of using bootstrap javascript rather than vanilla js, or whatever comes with choices. If there's no other way to do it tho....

Reply:
It is possible to handle the dropdown logic as part of the component, but it would be more complex. It'd look similar to [this](https://github.com/SleeplessOne1917/5E-Character-Sheet-Builder/blob/create-class/src/components/Select/Select/Select.tsx). 

The library that bootstrap uses for dropdowns, [Popper](https://popper.js.org/), is only 3kb gzipped (compared to choices.js, which is 19kb), so if you're worried about the bootstrap dependency making bundle sizes much larger, that shouldn't be an issue.